### PR TITLE
Chemical Manipulator - A more advanced chem master

### PR DESCRIPTION
**Pull request by Reviire**
Functionally identical to the chem master, but rather than clearing reagents when the beaker is ejected, it retains those reagents. It also has a large internal reagent storage to accommodate all the reagents you will be storing. 

It currently has the same requirements as the regular chem master, but you do need R&D to give you the stuff, or the mechanics. Otherwise, it should be working. Although the UI looks a bit different, I didn't really check what the other UI looks like.

I changed up the default chem master code to make implementing this easier, so there's some UI differences, although I'm not sure how large. Personally I don't think it's an issue, it still conveys the same information. 

:cl:
 * rscadd: Adds the Chemical Manipulator. Functionally identical to the chem master, but can hold reagents without a beaker with it's large internal storage.

### CODE REVIEW

**Comment by Reviire** [comment]
I don't understand github at all those commits weren't supposed to be there I was confused.

File: code/game/machinery/kitchen/snackbarmachine.dm [hunk 1]
```diff
@@ -6,6 +6,22 @@
...
 
+	var/max_snack_size = 10
+
+/obj/machinery/chem_master/RefreshParts()
+	var/scancount = 0
+	var/lasercount = 0
+	var/manipcount = 0
+
+	for(var/obj/item/weapon/stock_parts/SP in component_parts)
+		if(istype(SP, /obj/item/weapon/stock_parts/manipulator))
```
Comment (ShiftyRail) [inline code/game/machinery/kitchen/snackbarmachine.dm]: Make a list of all those types and use `if (is_type_in_list(SP, L)`.

File: code/game/machinery/kitchen/snackbarmachine.dm [hunk 2]
```diff
@@ -37,39 +54,85 @@
...
+	else
+		dat += "No beaker inserted."
+
+	// Buffer - Like normal chem masters, except retains without a beaker.
+	// Makes pills. Can flush or move to internal storage
+	dat += "<HR>"
+	dat += "<b>&ltInternal Snack Buffer&gt</b> <BR>"
+
+	var/b_mode_text
+	if(buffer_mode == 1)
```
Comment (ShiftyRail) [inline code/game/machinery/kitchen/snackbarmachine.dm]: Use a define for this.
`#define STORAGE 1` and then `if (buffer_mode == STORAGE)`.
Same for next line.

Additionally, you could use a switch statement.

File: code/modules/reagents/machinery/chem_master.dm [hunk 3]
```diff
@@ -136,418 +131,418 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
-			else if(href_list["percent"])
-				amount = reagents.get_reagent_amount(id) * text2num(href_list["percent"]) / 100
 			if(isnull(amount) || amount < 0)
 				return
-			if(mode)
-				reagents.trans_id_to(beaker, id, amount)
-			else
-				reagents.remove_reagent(id, amount)
+
+			if(storage_mode == 1)
```
Comment (ShiftyRail) [inline code/modules/reagents/machinery/chem_master.dm]: You could use defines as well as a switch statement here and in the following lines.

File: code/game/machinery/kitchen/snackbarmachine.dm [hunk 4]
```diff
@@ -6,6 +6,22 @@
 	chem_board = /obj/item/weapon/circuitboard/snackbar_machine
 	windowtype = "snackbar_machine"
 
+	var/max_snack_size = 10
+
+/obj/machinery/chem_master/RefreshParts()
```
Comment (MadmanMartian) [inline code/game/machinery/kitchen/snackbarmachine.dm]: You don't seem to be doing anything with these scancounts, lasercounts, and manipcounts. Why is this here if the max_snack_size modifier is commented out?

Comment (Reviire) [reply code/game/machinery/kitchen/snackbarmachine.dm]: It wasn't even named right, I dunno i just got sidetracked and started doing something, then got sidetracked again. It was pointless and I removed it.

**Comment by Reviire** [comment]
So it seems like everything is working as it should. It's always possible I'm retarded and I missed something big, but I think everything is working. Just need to get that sprite.

**Comment by Probe1** [comment]
CL and the PR needs a way better description of what this actually does.  "Way better" isn't descriptive.

**Comment by Probe1** [comment]
Reading all these formatting line changes in the middle of a PR that adds content gives me ptsd from  @dylanstrategie coding practices.

**Review by Probe1 — CHANGES_REQUESTED** [review]
Fix the ? on the machine please.  Otherwise I don't see any issues jumping out at me.

**Review by Probe1 — DISMISSED** [review]

File: code/modules/reagents/machinery/chem_master.dm [hunk 5]
```diff
@@ -3,8 +3,14 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
 
+#define BEAKER "Beaker"
+#define STORAGE "Storage"
+#define BUFFER "Buffer"
+#define FLUSH "FLUSH"
+
+
 /obj/machinery/chem_master
-	name = "\improper ChemMaster 3000"
+	name = "\improper Chem Master"
```
Comment (Probe1) [inline code/modules/reagents/machinery/chem_master.dm]: Why change this?

File: code/modules/reagents/machinery/chem_master.dm [hunk 6]
```diff
@@ -67,18 +81,21 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 	var/scancount = 0
 	var/lasercount = 0
 	var/manipcount = 0
+
 	for(var/obj/item/weapon/stock_parts/SP in component_parts)
 		if(istype(SP, /obj/item/weapon/stock_parts/manipulator))
-			manipcount += SP.rating-1
+			manipcount += SP.rating
```
Comment (Probe1) [inline code/modules/reagents/machinery/chem_master.dm:84]: Why change this?

File: code/game/machinery/kitchen/snackbarmachine.dm [hunk 7]
```diff
@@ -37,39 +39,80 @@
...
-	else
+	// Beaker
+	if(beaker)
 		var/datum/reagents/R = beaker.reagents
-		dat += "<A href='?src=\ref[src];eject=1'>Eject beaker and Clear Buffer</A><BR>"
-		if(!R.total_volume)
-			dat += "Beaker is empty."
-		else
-			dat += "Add to buffer:<BR>"
+		dat += "<A href='?src=\ref[src];eject=1'>Eject beaker.</A><BR>"
```
Comment (DamianX) [inline code/game/machinery/kitchen/snackbarmachine.dm:45]: remove this period

File: code/game/machinery/kitchen/snackbarmachine.dm [hunk 8]
```diff
@@ -37,39 +39,80 @@
...
+
+		if(R.total_volume)
+			// Beaker buttons
+			dat += {"
+				<table>
+					<td class="column1">
```
Comment (DamianX) [inline code/game/machinery/kitchen/snackbarmachine.dm:51]: Why is this class defined? Why is this even a table in the first place?

File: code/modules/reagents/machinery/chem_master.dm [hunk 9]
```diff
@@ -136,418 +136,468 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
+/obj/machinery/chem_master/attack_hand(mob/user as mob)
+	. = ..()
+	if(.)
+		return
+	user.set_machine(src)
+
+	var/dat = list()
+	// Pill bottle stuff
+	if(loaded_pill_bottle)
+		dat += "<A href='?src=\ref[src];ejectbottle=1'>Eject pill bottle.</A><BR>"
```
Comment (DamianX) [inline code/modules/reagents/machinery/chem_master.dm]: delete this period

File: code/modules/reagents/machinery/chem_master.dm [hunk 10]
```diff
@@ -136,418 +136,468 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
+	// Beaker
+	if(beaker)
+		var/datum/reagents/R = beaker.reagents
+		dat += "<A href='?src=\ref[src];eject=1'>Eject beaker.</A><BR>"
```
Comment (DamianX) [inline code/modules/reagents/machinery/chem_master.dm]: delete this period

File: code/modules/reagents/machinery/chem_master.dm [hunk 11]
```diff
@@ -136,418 +136,468 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
+
+		if(R.total_volume)
+			// Beaker buttons
+			dat += {"
+				<table>
```
Comment (DamianX) [inline code/modules/reagents/machinery/chem_master.dm:173]: why is this a table?

File: code/modules/reagents/machinery/chem_master.dm [hunk 12]
```diff
@@ -136,418 +136,468 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
-			src.Topic(null, list("amount" = "[useramount]", "removeall" = "1"))
-			return 1
+			src.Topic(null, list("amount" = "[useramount]", "storage_add" = "[id]"))
 
-		else if(href_list["toggle"])
-			mode = !mode
 			src.updateUsrDialog()
 			return 1
 
-		else if(href_list["main"])
```
Comment (DamianX) [inline code/modules/reagents/machinery/chem_master.dm:276]: you deleted this, so the "(Back)" button from the "(?)" won't work

File: code/modules/reagents/machinery/chem_master.dm [hunk 13]
```diff
@@ -136,418 +136,468 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
-						</a>"}
-				if (i%10 == 0)
-					dat +="<br>"
-			dat += "</div>"
-			//
-			// BUTTONS
-			//
-			dat += {"<HR><A href='?src=\ref[src];createpill=1'>Create single pill ([max_pill_size] units max)</A><BR>
-					<A href='?src=\ref[src];createpill_multiple=1'>Create multiple pills ([max_pill_size] units max each; [max_pill_count] max)</A><BR>
-					<A href='?src=\ref[src];createpill_multiple=1;createempty=1'>Create empty pills</A><BR>
```
Comment (DamianX) [inline code/modules/reagents/machinery/chem_master.dm:540]: bring this back, no reason to remove this, especially on the old chem master that you shouldn't be changing anyway

**Review by DamianX — CHANGES_REQUESTED** [review]

**Comment by DamianX** [comment]
image

**Comment by nuklearcellphoneg** [comment]
no pls finish this

**Pull request closed without merging**